src/cdromfs.c

```diff
@@ -248,9 +263,94 @@ struct cdrom_volume * cdrom_volume_open( int unit )
 	return 0;		
 }
 
-void cdrom_volume_close( struct cdrom_volume *v )
+static int cdrom_volume_close( struct volume *v )
 {
-	console_printf("cdromfs: umounted filesystem from unit %d\n",v->unit);
+	struct cdrom_volume *cdv = v->private_data;
+	console_printf("cdromfs: umounted filesystem from unit %d\n",cdv->unit);
 	kfree(v);
+	return 0;
 }
 
+static int cdrom_file_read(struct file *f, char *buffer, uint32_t n)
+{
+	struct cdrom_file *cdf = f->private_data;
+	struct cdrom_dirent *cdd = cdf->dirent;
+	uint32_t read = 0;
+	while (n > 0 && cdf->block * ATAPI_BLOCKSIZE + cdf->offset < cdd->length) {
+		uint32_t to_read = 0;
+		if (cdf->offset == cdf->buffer_size) {
+			cdrom_dirent_read_block(cdd, cdf->buffer, cdf->block);
+			if (cdd->length <= (cdf->block + 1) * ATAPI_BLOCKSIZE) {
+				cdf->buffer_size = cdd->length - cdf->block * ATAPI_BLOCKSIZE;
+			} else {
+				cdf->buffer_size = ATAPI_BLOCKSIZE;
+			}
+			cdf->block++;
+			cdf->offset = 0;
+		}
+		to_read = cdf->buffer_size - cdf->offset;
+		if (n < cdf->buffer_size - cdf->offset) {
+			to_read = n;
+		}
+		memcpy(buffer + read, cdf->buffer + cdf->offset, to_read);
+		cdf->offset += to_read;
+		n -= to_read;
+		read += to_read;
+	}
+	return read;
+}
+
+static struct file *cdrom_file_open(struct dirent *d, int8_t mode)
```

Comment:
I'm not sure what I'm supposed to check here. From what I understand, when opening a file, we should check:
- The file exists.
- The file can be read/written (depending on the user and what mode it was opened in).

But we already know the file exists because we have a `dirent`, and we don't really have the concept of modes or permissions yet. All we're really doing is allocating the struct that would contain the file data as we start to read it. Do I check that the pointer was correctly allocated, or just create the modes and check that we aren't trying to write to the file?

Reply:
Oh, I see.  yes, it seems that if you already have a dirent, then success is guaranteed.